feat(wiki): add an introduction to creating jobs (#45)
OpenJobDescription/openjd-specifications: wiki/Introduction-to-Creating-a-Job.md

# Introduction to Creating a Job

> This page guides you through developing a Job using Open Job Description. We'll explain elements of the Job Template bit by bit

Reviewer: First paragraph should be rewritten to explicitly call out the topics we're covering below, ideally with what we expect them to be web search'd by. So make sure to mention OpenJD here, Path Mapping, Chunking, etc.

as we walk through creating a Job that renders an animation with Blender and then encodes those frames into a video for viewing.
We'll also introduce you to some of the tooling that is available to help you rapidly develop a Job locally on your workstation.

If you have not already read through [How Jobs Are Constructed](How-Jobs-Are-Constructed) and [How Jobs Are Run](How-Jobs-Are-Run)

> then we recommend starting with those first and then coming back once complete. Those pages provide context that will help you

Reviewer: ```suggestion
then we recommend starting with those first. Those pages provide context that will help you
```

understand the remainder of this guide.

Table of Contents:

1. [Toolchain Setup](#1-toolchain-setup)
2. [Walkthrough](#2-walkthrough)
    1. [Starting with shell scripts](#21-starting-with-shell-scripts)
        1. [Creating a shell script to render with Blender](#211-creating-a-shell-script-to-render-with-blender)
        2. [Creating a shell script to encode a video with FFmpeg](#212-creating-a-shell-script-to-encode-a-video-with-ffmpeg)
    2. [Creating a Job Template](#22-creating-a-job-template)
        1. [Embed the scripts into a Job Template](#221-embed-the-scripts-into-a-job-template)
        2. [Parameterizing the template](#222-parameterizing-the-template)
        3. [Path Mapping](#223-path-mapping)
        4. [Adding Task Parallelism](#224-adding-task-parallelism)
    3. [Ready for Production](#23-ready-for-production)
3. [Next Steps](#3-next-steps)

> ## 1. Toolchain Setup

Reviewer: You need to mention that it requires Blender and FFmpeg installed as well.

Reviewer: I would fix this before merging

To replicate the steps taken in this guide you will need to have the [Open Job Description CLI](https://pypi.org/project/openjd-cli/)
installed. This CLI provides easy to use subcommands for validating the syntax of a Job Template, running Tasks defined by a Job
Template locally on your workstation, and more. The CLI is written in Python, so you will need to have Python 3.9 or higher
available on your workstation; for information on how to install Python, please see the official [Python.org website](https://www.python.org/).

> First, we suggest creating a [Python virtual environment](https://docs.python.org/3/library/venv.html) and installing the CLI into that:
> 
> | POSIX | Windows (CMD Shell) |
> | ----- | ------- | 
> | <pre>python3 -m venv .venv<br>source .venv/bin/activate<br>pip install openjd-cli</pre> | <pre>python3 -m venv .venv<br>.venv\Scripts\activate.bat<br>pip install openjd-cli</pre> | 

Reviewer: Out of scope.

```suggestion
We suggest installing the tooling into a [Python virtual environment](https://docs.python.org/3/library/venv.html).
```

For writing Job Templates by hand we also recommend using [Visual Studio Code](https://code.visualstudio.com/)

> and configuring it so that it can auto-complete the syntax of your Job Templates. To configure auto-complete, you

Reviewer: This is so cool!!!!

modify your [Visual Studio Code settings](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings)
to apply the Open Job Description schemas to your Job and Environment Templates. The following is an example of settings to
automatically apply the schemas to all JSON and YAML files on a workstation:

```json
    "json.schemas": [
        {"fileMatch": ["/**/*.json"], "url": "/Users/<username>/openjd-schemas/openjobdescription-jobtemplate-2023-09.json"},
        {"fileMatch": ["/**/*.json"], "url": "/Users/<username>/openjd-schemas/openjobdescription-environment-2023-09.json"}           
    ],
    "yaml.schemas": {
        "/Users/<username>/json-schemas/openjobdescription-jobtemplate-2023-09.json": [
            "/**/*.yaml"
        ],
        "/Users/myusername/json-schemas/openjobdescription-environment-2023-09.json": [
            "/**/*.yaml"
        ]
    }
```

To use auto-complete, simply use Control-Space and select a schema entity to automatically populate all of the entities required
fields, or select a field to auto-complete it:

| Menu | Selecting "JobTemplate" Entity | Selecting "specificationVersion" field |
| ---  | --- | ---- |

> | ![Autocomplete menu](images/template_autocomplete_menu.png) | ![Autocomplete template](images/template_autocomplete_entity.png) | ![Autocomplete field](images/template_autocomplete_field.png) |

Reviewer: Why are we still getting red underline errors with the specification installed?

Author: Mostly that the values are blank, so vs-code is highlighting that. The highlight on the standalone specificationVersion line is because there are other required fields that aren't yet present.

Note that if you prefer to programmatically generate Open Job Description templates, in a pipeline for instance,
then the [openjd-model](https://pypi.org/project/openjd-model/) Python package 
[supports that use case](https://github.com/OpenJobDescription/openjd-model-for-python?tab=readme-ov-file#converting-a-template-model-to-a-dictionary). 

## 2. Walkthrough

To demonstrate one way that you could develop a Job for Open Job Description this walkthrough will guide
you through creating a Job Template that generates frames of an animation using [Blender](https://www.blender.org/),
and then encodes those rendered frames into an mp4 video using [FFmpeg](https://ffmpeg.org/). We illustrate 
an approach to developing a Job that aims to keep your inner development loop fast by leveraging
the Open Job Description CLI to test the Job on your workstation. We also demonstrates making small incremental changes to make
it easier to narrow in on the root cause of an error. This process has faster iterations than submitting the Job

> to your distributed compute cluster, can be done in a cafe or airplane without an internet connection, and does not even require

Reviewer: ```suggestion
to your distributed compute cluster (AKA render farm), can be done in a cafe or airplane without an internet connection, and does not even require
```

that you have a compute cluster that supports Open Job Description.

### 2.1. Starting with shell scripts

> An Open Job Description Job is, essentially, describing a set of command-line commands to run, the inputs to those commands,

Reviewer: ```suggestion
An Open Job Description Job is, essentially, describing a set of command-line instructions to run, the inputs to those commands,
```

and the order in which they need to be run. We encourage you to think of authoring a Job as writing a series of shell
scripts that accomplish the objective of the Job when they are run in the correct order with the correct inputs.

> In this guide, you are creating a Job that generates frames of an animation using Blender and then encodes the frames into an animation
> using FFmpeg. You'll break this Job down into two steps:

Reviewer: ```suggestion
In this guide, you are creating a Job that generates a sequence of frames of a Blender animation and then encodes the frames into a movie file
using FFmpeg. You'll break this Job down into two steps:
```

> 1. Generate the animation frames with Blender; and then

Reviewer: ```suggestion
1. Generate the animation frames with Blender's command-line interface; and then
```

> 2. Encode the generated frames into an mp4 video with FFmpeg.

Reviewer: ```suggestion
2. Encode the generated frames into an mp4 video file with FFmpeg.
```

Note that this walkthrough is written for a bash-compatible shell such as those available on typical Linux or

> MacOS workstations. If you are using Windows, then we trust that the modifications to powershell or batch from bash
> will be straightforward enough for you to handle; though you may also run bash scripts on Windows if you install
> an application such as [Git BASH](https://gitforwindows.org/).

Reviewer: ```suggestion
MacOS workstations. If you are using Windows, you will need to modify the commands for PowerShell or batch; though you may also run bash scripts on Windows if you install
an application such as [Git BASH](https://gitforwindows.org/).
```

#### 2.1.1. Creating a shell script to render with Blender

> First, you'll create a shell script that runs Blender to render a part of the animation.

Reviewer: ```suggestion
First, you'll create a shell script that calls the Blender command-line interface to render a part of the animation.
```

Looking up the [documentation for the Blender command line](https://docs.blender.org/manual/en/latest/advanced/command_line/arguments.html)
we created the following as a starting point for our Blender script:

```bash
#!/bin/bash

# Return an error code if any command in the script fails.
set -eou pipefail

# Use Blender's scripting interface to reduce the scene resolution and sampling rate to speed up testing.

> # Remove this after testing is complete.

Reviewer: ```suggestion
# Remove this and the lines below, `  --python "$TMPFILE" \` and `rm -f $TMPFILE`, after testing is complete.
```

TMPFILE=$(mktemp)
cat > "$TMPFILE" << EOF
import bpy
for s in bpy.data.scenes:
  s.render.resolution_x = 480
  s.render.resolution_y = 270
bpy.context.scene.cycles.samples = 100
EOF

# Run blender to generate the image frames.
blender --background 3d/pavillon_barcelone_v1.2.blend \
  --python "$TMPFILE" \
  --render-output output_frames/frame-### \
  --render-format PNG --use-extension 1 \
  --render-frame "1..2"

# Cleanup the temporary python script
rm -f $TMPFILE
```

This script makes some assumptions, such as being run from a directory that contains our input and output, that we'll modify next, but
this is a good start. It uses Blender to render two frames of the 380 frame animation created by

> the ["Barcelona Pavillion" demo scene](https://www.blender.org/download/demo-files/#cycles) (accessed July 2024) at reduced resolution.

Reviewer: ```suggestion
the ["Barcelona Pavillion" demo scene](https://www.blender.org/download/demo-files/#cycles) by eMirage, distributed under CC-BY (accessed July 2024).
```

Reduced res has already been explained above. While we don't HAVE to attribute here (since we're not displaying the results of the file or using it directly, we should play it safe.

Save this to a file called `render.sh`, set its execute bit (`chmod +x render.sh`), and then test it to verify that it runs as expected:

```bash
% ./render.sh
Blender 4.1.1 (hash e1743a0317bc built 2024-04-16 00:06:22)
...
Fra:1 Mem:130.47M (Peak 206.18M) | Time:00:07.56 | Compositing | De-initializing execution
Saved: 'output_frames/frame-001.png'
...
Fra:2 Mem:130.48M (Peak 272.58M) | Time:00:07.64 | Compositing | De-initializing execution
Saved: 'output_frames/frame-002.png'
Time: 00:07.71 (Saving: 00:00.07)
```

Next, you'll want to be able to create a Job that can render any scene rather than just this one demo scene. So, parameterize the script
such that the scene file, output directory, and frames to render can all be passed in as arguments to the script:

```bash
#!/bin/bash

# Return an error code if any command in the script fails.
set -eou pipefail

TMPFILE=$(mktemp)
cat > "$TMPFILE" << EOF
import bpy
for s in bpy.data.scenes:
  s.render.resolution_x = 480
  s.render.resolution_y = 270
bpy.context.scene.cycles.samples = 100
EOF

> SCENE="$1"

Reviewer: Recommend added a comment line above this which quickly explains that `$1` etc will translate into arguments on the command line calling the script, with the script itself being `$0`

```bash
% ./render.sh 3d/pavillon_barcelone_v1.2.blend output_frames 1 2
Blender 4.1.1 (hash e1743a0317bc built 2024-04-16 00:06:22)
...
Fra:2 Mem:130.48M (Peak 272.58M) | Time:00:08.09 | Compositing | De-initializing execution
Saved: 'output_frames/frame-002.png'
Time: 00:08.17 (Saving: 00:00.07)
```

#### 2.1.2. Creating a shell script to encode a video with FFmpeg

To create the script for FFmpeg encoding consult the [Academy Software Foundation's encoding guidelines](https://academysoftwarefoundation.github.io/EncodingGuidelines/) on how to run FFmpeg:

```bash
#!/bin/bash

set -eou pipefail

ffmpeg -y -r 10 -start_number 1 -i output_frames/frame-%03d.png -pix_fmt yuv420p \
    -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
    -frames:v 300 -c:v libx264 -preset fast \
    -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
    -movflags faststart "animation.mp4"
```

Then, as with the Blender script, save it to a file (`encode.sh`), set its execute bits, and then test it:

```bash
% ./encode.sh 
ffmpeg version 6.1.1-tessus  https://evermeet.cx/ffmpeg/  Copyright (c) 2000-2023 the FFmpeg developers
...
```

> Then modify it so that it accepts script arguments:

Reviewer: script arguments or CLI arguments?

Author: command-line args. Clarified in doc

and the processes that will run.

### 2.2.1. Embed the scripts into a Job Template

Starting with a skeleton of the minimal required properties of a Job Template:

```yaml
specificationVersion: jobtemplate-2023-09
name: 
steps:
  - name: 
    script:
      actions:
        onRun:
          command: 
```

Fill in the names for the Job that you're going to create, and the first Step that will run the Blender
rendering script:

```yaml
specificationVersion: jobtemplate-2023-09
name: DemoJob
steps:
  - name: BlenderRender
    script:
      actions:
        onRun:
          command: 
```

The way to think about a Step in Open Job Description is that it is defining a command to run and the collection of inputs
to run the command with. This is a [map operation](https://en.wikipedia.org/wiki/Map_(higher-order_function)), if
you are familiar with that concept. You're not going to define any inputs for the command yet, so you'll be using the default
of having a single Task with empty input. The command is defined in the `onRun` Action in the skeleton above; this is
the command that is run when running a Task within a [Session](How-Jobs-Are-Run#sessions). An Action in the 2023-09 

> revision of the specification [is defined as](2023-09-Template-Schemas#5-action):

Reviewer: We suddenly up the reading and tech level a lot with this paragraph and the few that follow. Consider shorter sentences and more of them.

Author: I've adjusted this significantly. Let me know what you think.

```yaml
command: <CommandString> # @fmtstring[host]
args: [ <ArgString>, ... ] # @optional @fmtstring[host]
timeout: <posinteger> # @optional
cancelation: <CancelationMethod> # @optional
```

Notice that both the `command` and elements of the `args` array are annotated with `@fmtstring[host]` indicating that
they are [Format Strings](How-Jobs-Are-Constructed#format-strings) that are evaluated on the host that the command will
run on. We're going to make use of this property to create our Job.

> You've already expressed your Job as a series of two shell scripts to run in sequence, so next you'll make use of the

Reviewer: Feels like we skipped a step- we haven't had the OpenJD example of using the script itself yet, so "you've already expressed your job as.." doesn't make much sense. We've only been running single scripts, not any Jobs yet

Reviewer: Coming back with fresh eyes, I would phrase it as "We've created two scripts and can manually run them in sequence to approximate the results we want. We can use Open Job Description's concept of embedded files to put those scripts directly into a Job template, which..

[Embedded File](2023-09-Template-Schemas#6-embeddedfile) concept to make those scripts available to the Job. An embedded file
lets you define the contents of a text file directly in a Job Template, then that text file will be created in the 
[Session's temporary Working Directory](How-Jobs-Are-Run#sessions) when the Task is run.

The result is the start of your Job Template, which we save to file called `job.template.yaml`:

```yaml
specificationVersion: jobtemplate-2023-09
name: DemoJob
steps:
  - name: BlenderRender
    script:
      actions:
        onRun:
          # Note: Task.File.Render refers to the file location on disk where the
          #  contents of the embedded file named "Render" are materialized.
          command: "{{Task.File.Render}}"

>           args:

Reviewer: Be good to comment here that this *needs to be a list* as it's one area we've been people butt their heads against (including me)

Author: Added something.

            - 3d/pavillon_barcelone_v1.2.blend
            - output_frames
            - "1"
            - "2"
      embeddedFiles:
        - name: Render
          type: TEXT
          filename: render.sh
          runnable: true
          data: |
            #!/bin/bash

            # Return an error code if any command in the script fails.
            set -eou pipefail

>             TMPFILE=$(mktemp)

Reviewer: This is missing the comments from the above

            cat > "$TMPFILE" << EOF
            import bpy
            for s in bpy.data.scenes:
              s.render.resolution_x = 480
              s.render.resolution_y = 270
            bpy.context.scene.cycles.samples = 100
            EOF

            SCENE="$1"
            OUTDIR="$2"
            START_FRAME="$3"
            END_FRAME="$4"

            blender --background "$SCENE" \
              --python "$TMPFILE" \
              --render-output "$OUTDIR"/frame-### \
              --render-format PNG --use-extension 1 \
              --render-frame "${START_FRAME}..${END_FRAME}"

            rm -f $TMPFILE
```

> Note that if you are writing your Job for Windows with the render script being a batch or powershell script, then you will

Reviewer: ```suggestion
Note that if you are writing your Job for Windows with the render script being a batch or PowerShell script, then you will
```

need to make a couple of changes from the above:

1. Change the `filename` in the embedded file to have the correct filename suffix (`.bat` for batch or `.ps1` for powershell); and
2. For powershell, change the `onRun` action to:

```yaml
command: powershell
args:
  - "-File"
  - "{{Task.File.Render}}"
  - 3d/pavillon_barcelone_v1.2.blend
  - output_frames
  - "1"
  - "2"
```

Next, following our incremental approach, test what you have so far locally on your workstation using the
Open Job Description CLI (reminder: activate your Python virtual environment first, if you are using one). First,

> make sure that you do not have any syntax errors in our template using the CLI's `check` command:

Reviewer: ```suggestion
Before moving on, it's a good idea to check to the syntax of the file. Make sure to activate any Python virtual env as needed, then use the CLI's `check` command:
```

```bash
% openjd check job.template.yaml
Template at 'job.template.yaml' passes validation checks.
```

Then, you can run it using the CLI's `run` command to make sure that it runs:

```bash
% openjd run --step BlenderRender job.template.yaml
Fri Jul  5 09:33:02 2024	
Fri Jul  5 09:33:02 2024	==============================================
Fri Jul  5 09:33:02 2024	--------- Running Task
Fri Jul  5 09:33:02 2024	==============================================
...
Fri Jul  5 09:33:02 2024	----------------------------------------------
Fri Jul  5 09:33:02 2024	Phase: Running action
Fri Jul  5 09:33:02 2024	----------------------------------------------

> Fri Jul  5 09:33:02 2024	Running command /private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_session0bd5l1au/tmphe6bjl3c.sh

Reviewer: I always simplify the GUIDs when giving examples to keep the line length down- so `/private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/` becomes something like `/private/var/folders/93/p7gr/`

Fri Jul  5 09:33:02 2024	Command started as pid: 85706
Fri Jul  5 09:33:02 2024	Output:
Fri Jul  5 09:33:03 2024	Blender 4.1.1 (hash e1743a0317bc built 2024-04-16 00:06:22)
Fri Jul  5 09:33:03 2024	Error: Cannot read file "/private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_session0bd5l1au/3d/pavillon_barcelone_v1.2.blend": No such file or directory
Fri Jul  5 09:33:03 2024	
Fri Jul  5 09:33:03 2024	Blender quit
...
```

> Uh-oh, there's an error! Blender couldn't find the input file. This is because all actions in a Session are always run with their

Reviewer: Nit: "The error means that Blender couldn't find the input file."

current working directory being the Session's temporary working directory

> (`/private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_session0bd5l1au` in this case), the scene file argument to your

Reviewer: ```suggestion
current working directory being the Session's temporary working directory, which was `/private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_session0bd5l1au` in this case. In our script, we've been using a relative path...
```

          command: "{{Task.File.Render}}"
          args:
            - /Users/myusername/blender_demo/3d/pavillon_barcelone_v1.2.blend
            - /Users/myusername/blender_demo/output_frames
            - "1"
            - "2"
...
```

Then check the template syntax and try to run it again:

```bash
% openjd check job.template.yaml            
Template at 'job.template.yaml' passes validation checks.
% openjd run --step BlenderRender job.template.yaml
Fri Jul  5 09:42:01 2024	
Fri Jul  5 09:42:01 2024	==============================================
Fri Jul  5 09:42:01 2024	--------- Running Task
Fri Jul  5 09:42:01 2024	==============================================
...
Fri Jul  5 09:42:18 2024	Open Job Description CLI: All actions completed successfully!
Fri Jul  5 09:42:18 2024	Open Job Description CLI: Local Session ended! Now cleaning up Session resources.

--- Results of local session ---

Session ended successfully

Job: DemoJob
Step: BlenderRender
Duration: 17.484004959 seconds
Tasks run: 1
```

Success! Also verify that the output are where we expect them to be:

```bash
% ls /Users/myusername/blender_demo/output_frames 
frame-001.png	frame-002.png
```

> Before we continue adding functionality to your Job we're going to have you add a timeout to the Task. The timeout

Author: Bug to fix, ideally before this merges: https://github.com/OpenJobDescription/openjd-cli/issues/96

property of an action limits how long that action is allowed to run. When the time limit is reached then the
command that's running will be canceled using the action's defined [cancelation method](2023-09-Template-Schemas#53-cancelationmethod)
(by default, all processes in the Task's process tree are stopped). The timeout is a guard-rail to help prevent misbehaving actions

> from running forever; which will incur a monetary cost if your compute is, say, running on a cloud provider. Add a short timeout,

Reviewer: ```suggestion
from running forever; which can incur direct costs depending on your infrastructure (such as a cloud provider).

To test the timeout functionality works as expected, add a too short timeout,
```

test that it works using the CLI, and then modify the timeout value to something large enough to complete the Task:

```yaml
...
        onRun:
          command: "{{Task.File.Render}}"
          args:
            - /Users/myusername/blender_demo/3d/pavillon_barcelone_v1.2.blend
            - /Users/myusername/blender_demo/output_frames
            - "1"
            - "2"
          timeout: 2
...
```

Then run the change:

```bash
% openjd run --step BlenderRender job.template.yaml
Fri Jul  5 09:59:28 2024	
Fri Jul  5 09:59:28 2024	==============================================
Fri Jul  5 09:59:28 2024	--------- Running Task
Fri Jul  5 09:59:28 2024	==============================================
...
Fri Jul  5 09:59:30 2024	TIMEOUT - Runtime limit reached at 2024-07-05T14:59:30Z. Canceling action.
Fri Jul  5 09:59:30 2024	Canceling subprocess 88373 via termination method at 2024-07-05T14:59:30Z.
...
```

> Finally, to complete this part of the guide add a Step that encodes the video to your Job. This will follow the same

Reviewer: ```suggestion
As expected, a timeout of 2 seconds is too short to allow the Blender render to complete, and the Job is cancelled. 

Make sure to modify the timeout to a larger value before going forward.

To complete this part of the guide, add a Step that encodes the video to your Job. This will follow the same
```

    script:
      actions:
        onRun:
          command: "{{Task.File.Encode}}"
          args:
            - /Users/myusername/blender_demo/output_frames
            - /Users/myusername/blender_demo/animation.mp4
            - "1"
          timeout: 60
      embeddedFiles:
        - name: Encode
          type: TEXT
          runnable: true
          filename: encode.sh
          data: |
            #!/bin/bash

            set -eou pipefail

            INPUT_DIR="$1"
            OUTPUT_FILENAME="$2"
            START_FRAME="$3"

            ffmpeg -y -r 10 -start_number "$START_FRAME" -i "$INPUT_DIR"/frame-%03d.png -pix_fmt yuv420p \
                -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
                -frames:v 300 -c:v libx264 -preset fast \
                -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
                -movflags faststart "$OUTPUT_FILENAME"

```

Notice that the EncodeVideo Step differs from the BlenderRender Step in that it includes the definition of the Step's
[dependencies](2023-09-Template-Schemas#32-stepdependency). This ensures that when you submit the Job to a compute cluster

> to run that the `BlenderRender` Step will run successfully to completion before the `EncodeVideo` Step is run.

Reviewer: ```suggestion
to run that the `BlenderRender` Step will run successfully to completion before the `EncodeVideo` Step is started. 
```

> As a final note, if your shell scripts in your actual Jobs are large then you may prefer to host them on a shared network fileshare

Reviewer: ```suggestion
If your shell scripts in your actual Jobs are large then you may prefer to host them on a shared network fileshare
```

Not really final, just the end of the part.

Reviewer: I also could take or leave this final paragraph

instead of embedding them directly within a Job Template. If you do so, then to avoid a potential command-injection attack on your
Jobs we remind you to ensure that no unauthorized users (including any running Job on the cluster) can possibly overwrite your scripts. 

### 2.2.2. Parameterizing the template

> At this point you have a Job Template that has a specific scene file and frame range hard-coded into it. You can make this more

Reviewer: Move the hard-coded bit up front to make it clear
```suggestion
At this point you have a Job Template that has hard-coded values for the scene file and frame range. You can make this more
```

general by adding [Job Parameters](2023-09-Template-Schemas#2-jobparameterdefinition) to the template to allow changing things
like the scene file, frame range, and output locations when using the template to create a job.

In the spirit of small incremental improvements, let's start by turning the scene file into a parameter. This is a filename, so
it will be a [`PATH` type job parameter](2023-09-Template-Schemas#22-jobpathparameterdefinition). Modify the template to include
the parameter definition:

```yaml
specificationVersion: jobtemplate-2023-09
name: DemoJob
parameterDefinitions:
  - name: SceneFile
    type: PATH
    dataFlow: IN
    objectType: FILE
steps:
...
```

Then make sure that the syntax is correct:

```bash
% openjd check job.template.yaml
Template at 'job.template.yaml' passes validation checks.
```

Next, if you run the updated template and provide a value for this parameter then the behavior of the Job will not change because we
are still hard-coding the location of the scene file within the template.

> 

Reviewer: I think this is a little tooooo step by step. It'd be more clear if you defined and referenced the parameter in a single step

Author: Thanks. Condensed.

```

So, further update the template to reference the job parameter named `SceneFile` instead of using the hard-coded scene:

```yaml
...
  - name: BlenderRender
    script:
      actions:
        onRun:
          command: "{{Task.File.Render}}"
          args:
            - "{{Param.SceneFile}}"
            - /Users/myusername/blender_demo/output_frames
            - "1"
            - "2"
...
```

Then check the syntax and run it:

```bash
% openjd check job.template.yaml
Template at 'job.template.yaml' passes validation checks.

# Test the failing case to make sure that our change took effect
% openjd run --step BlenderRender job.template.yaml -p SceneFile=/does/not/exist
...
Fri Jul  5 14:18:11 2024	----------------------------------------------
Fri Jul  5 14:18:11 2024	Phase: Running action
Fri Jul  5 14:18:11 2024	----------------------------------------------
Fri Jul  5 14:18:11 2024	Running command /private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_sessionvg8p3pkk/tmpend9rf4c.sh
Fri Jul  5 14:18:11 2024	Command started as pid: 24863
Fri Jul  5 14:18:11 2024	Output:
Fri Jul  5 14:18:12 2024	Blender 4.1.1 (hash e1743a0317bc built 2024-04-16 00:06:22)

> Fri Jul  5 14:18:12 2024	Error: Cannot read file "/does/not/exist": No such file or directory

Reviewer: and skip the error, just provide the value that used to be where `{{Param.SceneFile}}` is now.

...
Session ended with errors; see Task logs for details

Job: DemoJob
Step: BlenderRender
Duration: 0.647131209 seconds
Tasks run: 0

# And now the passing case
% openjd run --step BlenderRender job.template.yaml -p SceneFile=$(pwd)/3d/pavillon_barcelone_v1.2.blend
...
Session ended successfully

Job: DemoJob
Step: BlenderRender
Duration: 17.102627334 seconds
Tasks run: 1
```

Repeating the same for the other values that we want to parameterize:

1. The location of the `output_frames` directory and the output animation file's name are both [`PATH` type job parameters](2023-09-Template-Schemas#22-jobpathparameterdefinition);
2. The starting and ending frame number for the animation are both [`INT` type job parameters](2023-09-Template-Schemas#23-jobintparameterdefinition); and
3. The name of the job can be a [`STRING` type job parameter](2023-09-Template-Schemas#21-jobstringparameterdefinition).

> Pause here and try to update those yourself before continuing. The result will end up looking equivalent to:

Reviewer: ```suggestion
The result will end up looking equivalent to:
```

```yaml
specificationVersion: jobtemplate-2023-09
name: "{{Param.JobName}}"
parameterDefinitions:
  - name: SceneFile
    type: PATH
    dataFlow: IN
    objectType: FILE
  - name: FramesDirectory
    type: PATH
    dataFlow: OUT
    objectType: DIRECTORY
  - name: AnimationFile
    type: PATH
    dataFlow: OUT
    objectType: FILE
  - name: FrameStart
    type: INT
    minValue: 1
    default: 1
  - name: FrameEnd
    type: INT
    minValue: 1
    default: 2
  - name: JobName
    type: STRING
    minLength: 1
    default: "DemoJob"
steps:
  - name: BlenderRender
    script:
      actions:
        onRun:
          command: "{{Task.File.Render}}"
          args:
            - "{{Param.SceneFile}}"
            - "{{Param.FramesDirectory}}"
            - "{{Param.FrameStart}}"
            - "{{Param.FrameEnd}}"
          timeout: 60
      embeddedFiles:
        - name: Render
          type: TEXT
          filename: render.sh
          runnable: true
          data: |
            #!/bin/bash

            # Return an error code if any command in the script fails.
            set -eou pipefail

>             TMPFILE=$(mktemp)

Reviewer: Needs the comments

          filename: encode.sh
          data: |
            #!/bin/bash

            set -eou pipefail

            INPUT_DIR="$1"
            OUTPUT_FILENAME="$2"
            START_FRAME="$3"

            ffmpeg -y -r 10 -start_number "$START_FRAME" -i "$INPUT_DIR"/frame-%03d.png -pix_fmt yuv420p \
                -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
                -frames:v 300 -c:v libx264 -preset fast \
                -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
                -movflags faststart "$OUTPUT_FILENAME"
```

Finally, running it with the CLI:

```bash
% openjd run --step BlenderRender job.template.yaml \
  -p SceneFile=$(pwd)/3d/pavillon_barcelone_v1.2.blend \
  -p FramesDirectory=$(pwd)/output_frames \
  -p AnimationFile=$(pwd)/animation.mp4 \
  -p FrameStart=1 -p FrameEnd=2
Fri Jul  5 14:28:30 2024	
Fri Jul  5 14:28:30 2024	==============================================
Fri Jul  5 14:28:30 2024	--------- Running Task
Fri Jul  5 14:28:30 2024	==============================================
...
Session ended successfully

Job: DemoJob
Step: BlenderRender
Duration: 18.346753125 seconds
Tasks run: 1
```

> 

Reviewer: Note here that we're still running at low resolution for testing.

the operating systems might be the same but shared network filesystems may be mounted in different locations; or the assets reside
on the local hard drive on your workstation but will be automatically uploaded to an asset management system and materialized into
a different location on the compute host when the Job is running. For these kinds of cases, Open Job Description provides a
[path mapping](How-Jobs-Are-Run#path-mapping) mechanism to assist remapping the location of files directories from your workstation
to their location on the compute host where the Job will run.

The template that you've written so far is already set up to make use of the path mapping mechanism by using PATH type job parameters,
rather than STRING types for parameters that are files and directories. The PATH type parameters will automatically have any path mapping
rules applied to them when resolving the value in a template.

You can use the CLI to test the path mapping of this job template as follows. We, ultimately, want our job to use the files in our
current directory when running, but pretend that you're submitting from a different workstation where the files are located

> in the `/mnt/shared/demo` directory. First, create an environment variable that contains the path mapping rules

Reviewer: This isn't super clear and could use some rewording

Author: Reworded.

[according to the specification](How-Jobs-Are-Run#path-mapping) as though you are submitting from that fictional workstation and
that the Job will be run on your workstation:

```bash

> PATH_MAPPING_RULES="{\"version\":\"pathmapping-1.0\", \"path_mapping_rules\": [{\"source_path_format\": \"POSIX\", \"source_path\": \"/mnt/shared/demo\", \"destination_path\": \"$(pwd)\"}]}"

Reviewer: Before dumping this on the reader, it would be best to break it out in a more readable format and describe what literal mapping this path mapping json will accomplish

Author: Fixed up with the rewording above. Let me know what you think?

Fri Jul  5 17:09:47 2024	Running command /private/var/folders/93/p7mk60kn69ggyc87dfbwtht80000gr/T/OpenJD/sample_sessionbr6836aq/tmp3x4fnush.sh
Fri Jul  5 17:09:47 2024	Command started as pid: 41372
Fri Jul  5 17:09:47 2024	Output:
Fri Jul  5 17:09:48 2024	Blender 4.1.1 (hash e1743a0317bc built 2024-04-16 00:06:22)
Fri Jul  5 17:09:48 2024	Read prefs: "/Users/myusername/Library/Application Support/Blender/4.1/config/userpref.blend"
Fri Jul  5 17:09:48 2024	Read blend: "/Users/myusername/blender_demo/3d/pavillon_barcelone_v1.2.blend"
...
```

This works because the `pavillon_barcelone_v1.2.blend` file references the files that it needs (the textures beside it in the directory)
with relative path references. If, on the other hand, `pavillon_barcelone_v1.2.blend` referenced files with absolute paths then this would
not work as-is because Blender does not know where the texture files that it needs have been relocated to on the compute host. You can
demonstrate this by opening the scene in Blender, changing the external references to absolute 
(`File -> External Data -> Make paths absolute`), saving, and then moving the scene file and textures to a different directory and

> re-running the job. You'll notice that the resulting images are all pink due to missing textures.

Reviewer: This feels like a complication that might not be worth mentioning at the target reader level. Maybe just a quick aside about "The Blender file only contains internal or relative references"

Author: Yeah, it's a bit much.

the Step's definition. You use this property to define the collection of inputs that the Step's `onRun` action is run with.
In this case, you can create Tasks that either each render a single frame or each render a contiguous group of frames.
Let's go over both.

#### 2.2.4.1. One Frame per Task

The `render.sh` embedded script that you're using in the Job currently renders a sequence of frames rather than a single frame, so
to render a single frame in each Task you'll first need to modify and test the `render.sh` embedded script.
We suggest making the modifications outside of the template, testing them, and then copying the modifications into the template when
complete. This will help you make rapid changes without having to go through the Open Job Description CLI and let you benefit from
the features of your IDE, such as syntax highlighting and validation, that help you author shell scripts. The modified script to render
a single frame at a time looks like:

```bash
#!/bin/bash

# Return an error code if any command in the script fails.
set -eou pipefail

> TMPFILE=$(mktemp)

Reviewer: Copy the comments from above

Next, modify the command arguments for the Step to pass the value of `Frame` to the `render.sh` script to match our modification to it:

```yaml
...
      actions:
        onRun:
          command: "{{Task.File.Render}}"
          args:
            - "{{Param.SceneFile}}"
            - "{{Param.FramesDirectory}}"
            - "{{Task.Param.Frame}}"
          timeout: 60
...
```

That's it. You're ready to check the syntax of the template, and use the CLI to test the change locally by running a single Task
from the Step:

```bash
% openjd check job.template.yaml 
Template at 'job.template.yaml' passes validation checks.

% openjd run --step BlenderRender job.template.yaml \
  -tp Frame=100 \
  -p SceneFile=$(pwd)/3d/pavillon_barcelone_v1.2.blend \
  -p FramesDirectory=$(pwd)/output_frames \
  -p AnimationFile=$(pwd)/animation.mp4 \
  -p FrameStart=1 -p FrameEnd=380
Mon Jul  8 10:08:24 2024	
Mon Jul  8 10:08:24 2024	==============================================
Mon Jul  8 10:08:24 2024	--------- Running Task
Mon Jul  8 10:08:24 2024	==============================================
Mon Jul  8 10:08:24 2024	Parameter values:
Mon Jul  8 10:08:24 2024	Frame(INT) = 100
...
```

> When you run the CLI without the `-tp` argument, then you'll see all of the individual Tasks in the Step being run independently:

Reviewer: If you were to run the CLI without the argument.... but if you did, then it would try and render all 380 frames..

Author: Reworded this a bit.

            set -eou pipefail

            INPUT_DIR="$1"
            OUTPUT_FILENAME="$2"
            START_FRAME="$3"

            ffmpeg -y -r 10 -start_number "$START_FRAME" -i "$INPUT_DIR"/frame-%03d.png -pix_fmt yuv420p \
                -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
                -frames:v 300 -c:v libx264 -preset fast \
                -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
                -movflags faststart "$OUTPUT_FILENAME"

```

#### 2.2.4.2. Multiple Frames Per Task

Some rendering applications may take a significant amount of time to startup and load the data that it needs to work. In this case,
you may wish to render multiple contiguous frames of the animation in each Task. There are two ways that you can accomplish this
with Open Job Description:

1. Define Task parameters for the start and end of the range of frames to render in each Task and leverage a
   [combination expression](2023-09-Template-Schemas#343-combinationexpr) to combine them; or
2. Define Task parameter as a `STRING` type and hardcode the range of frames that we want to render in each Task.

We'll go through both options.

> ##### 2.2.4.2.1. With a Combination Expression

Author: This section needs the feature request https://github.com/OpenJobDescription/openjd-model-for-python/issues/131 to be implemented before it can be published.

the collection of Task values for the Step. For example, if you have a parameter space like:

```yaml
parameterSpace:
  taskParameterDefinitions:
  - name: Num
    type: INT
    range: "1-3"
  - name: Letter
    type: STRING
    range:
    - 'A'
    - 'B'
    - 'C'
```

Then the default combination expression `Num * Letter` produces a parameter space with the nine values 
`{Num=1, Letter=A}, {Num=1, Letter=B}, ... {Num=3, Letter=B}, {Num=3, Letter=C}`. However, a combination expression using the
association operator, `(Num,Letter)`, produces a parameter space with only three values:
`{Num=1,Letter=A}, {Num=2,Letter=B}, {Num=3,Letter=C}`.

Using the association operator, define the Step's parameter space as:

```yaml
...
    parameterSpace:
      taskParameterDefinitions:
        # RangeStart = 1,12,23,34,...,374
        - name: RangeStart
          type: INT
          range: "1-380:11"
        # RangeEnd = 11,22,33,44,...373,380
        - name: RangeEnd
          type: INT
          range: "11-380:11,380"
      # Tasks = (1,11), (12,22), (23,33), ... etc
      combination: "(RangeStart,RangeEnd)"
...
```

> 

Reviewer: Give the example of what the above will result in, and what the `:` is doing, as this is non-obvious. Because :11 results in an integer being returned for the range every 11 values.. etc.

Author: I've added an aside to help.

      combination: "(RangeStart,RangeEnd)"
    script:
      actions:
        onRun:
          command: "{{Task.File.Render}}"
          args:
            - "{{Param.SceneFile}}"
            - "{{Param.FramesDirectory}}"
            - "{{Task.Param.RangeStart}}..{{Task.Param.RangeEnd}}"
...
```

Then check the syntax for the template, and test it locally using the CLI:

```bash
% openjd check job.template.yaml                     
Template at 'job.template.yaml' passes validation checks.

% openjd run --step BlenderRender job.template.yaml \
  -p SceneFile=$(pwd)/3d/pavillon_barcelone_v1.2.blend \
  -p FramesDirectory=$(pwd)/output_frames \
  -p AnimationFile=$(pwd)/animation.mp4 \
  -p FrameStart=1 -p FrameEnd=380 -p FramesPerTask=11 \
  --tasks '[{"RangeStart": 1, "RangeEnd": 11}]'
Mon Jul  8 12:23:56 2024	
Mon Jul  8 12:23:56 2024	==============================================
Mon Jul  8 12:23:56 2024	--------- Running Task
Mon Jul  8 12:23:56 2024	==============================================
Mon Jul  8 12:23:56 2024	Parameter values:
Mon Jul  8 12:23:56 2024	RangeStart(INT) = 1
Mon Jul  8 12:23:56 2024	RangeEnd(INT) = 11
...
```

> The complete Job Template is:

Reviewer: ```suggestion
The Job Template is now:
```

            START_FRAME="$3"

            ffmpeg -y -r 10 -start_number "$START_FRAME" -i "$INPUT_DIR"/frame-%03d.png -pix_fmt yuv420p \
                -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
                -frames:v 300 -c:v libx264 -preset fast \
                -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
                -movflags faststart "$OUTPUT_FILENAME"
```

##### 2.2.4.2.2 As STRING Type

This option is only suitable for cases where you will be writing or programmatically generating a Job Template for a specific input
because the `FrameStart` and `FrameEnd` values need to be known when defining the set of values for the Task parameter. To create the 
parameter space, decide how many frames you want in each Task (we picked 40 in this case to cut down on typing) and then write out
the frame expressions that Blender requires as the values of the Task parameter

```yaml
parameterSpace:
      taskParameterDefinitions:
        - name: Frames
          type: STRING
          range:
          - "1..40"
          - "41..80"
          - "81..120"
... and so on until
          - "361-380"
...
```

> Also remove the definitions of the `FrameStart` and `FrameEnd` Job Parameters from the template because they are no longer used,

Reviewer: I don't think we need the rest of this example to be fully demonstrated. We can list the caveats and gotchas and move on.

            START_FRAME="$3"

            ffmpeg -y -r 10 -start_number "$START_FRAME" -i "$INPUT_DIR"/frame-%03d.png -pix_fmt yuv420p \
                -vf "scale=in_color_matrix=bt709:out_color_matrix=bt709" \
                -frames:v 300 -c:v libx264 -preset fast \
                -color_range tv -colorspace bt709 -color_primaries bt709 -color_trc iec61966-2-1 \
                -movflags faststart "$OUTPUT_FILENAME"
```

As usual, you can check the syntax of the template and then try it out locally using the CLI:

```bash
% openjd check job.template.yaml                     
Template at 'job.template.yaml' passes validation checks.

% openjd run --step BlenderRender job.template.yaml \
  -tp Frames=1..40 \
  -p SceneFile=$(pwd)/3d/pavillon_barcelone_v1.2.blend \
  -p FramesDirectory=$(pwd)/output_frames \
  -p AnimationFile=$(pwd)/animation.mp4
Mon Jul  8 10:48:30 2024	
Mon Jul  8 10:48:30 2024	==============================================
Mon Jul  8 10:48:30 2024	--------- Running Task
Mon Jul  8 10:48:30 2024	==============================================
Mon Jul  8 10:48:30 2024	Parameter values:
Mon Jul  8 10:48:30 2024	Frames(STRING) = 1..40
...
```

> ### 2.3. Ready for Production

Author: Note to self: We should add something on the `hostRequirements` property in here.

To develop the Job Template for this guide you've been reducing the resolution and quality of the resulting frames to make the
development iterations quicker. To use the template in production you'd either need to remove the testing code from the `render.sh`
embedded file so that it becomes:

```bash
#!/bin/bash

# Return an error code if any command in the script fails.
set -eou pipefail

SCENE="$1"
OUTDIR="$2"
FRAME="$3"

blender --background "$SCENE" \
  --render-output "$OUTDIR"/frame-### \
  --render-format PNG --use-extension 1 \
  --render-frame "${FRAME}"
```

> or you need to modify the template so that it has both a testing and production mode. We'll do the later in this guide to demonstrate

Reviewer: I think we should break out this guide is long enough to break out into multiple pages. If we do that, feel free to ignore some of my earlier comments are removals.

using [Environments](2023-09-Template-Schemas#4-environment) in a Job Template. Think of Environments as encapsulating commands that you
can run [before any Task has run, and after all Tasks have completed](How-Jobs-Are-Run#sessions) (whether successfully or not).
Since they're just arbitrary commands of your choosing they can do anything that you would like, but one of their use cases is to

> set up or modify the environment where your Tasks will be running. For this case, modify the `render.sh` embedded file
> so that the testing code is only run if the environment variable `TESTING_TEMPLATE` has the value `true`:

Reviewer: ```suggestion
set up or modify the environment where your Tasks will be running.

For this case, modify the `render.sh` embedded file
so that the testing code is only run if the environment variable `TESTING_TEMPLATE` has the value `true`:
```

```bash
#!/bin/bash

# Return an error code if any command in the script fails.
set -eou pipefail

TMPFILE=$(mktemp)
if test "${TESTING_TEMPLATE:-false}" == "true"; then
    cat > "$TMPFILE" << EOF
import bpy
for s in bpy.data.scenes:
s.render.resolution_x = 480
s.render.resolution_y = 270
bpy.context.scene.cycles.samples = 100
EOF
fi

SCENE="$1"
OUTDIR="$2"
FRAME="$3"

blender --background "$SCENE" \

>   --python "$TMPFILE" \

Reviewer: this won't error with an empty file?

Author: Nope. Turns out that an empty file is a perfectly valid python script ;-)
